src/org/zaproxy/zap/extension/pscan/scanner/RegexAutoTagScanner.java

```diff
@@ -310,9 +310,15 @@ public boolean equals(Object object) {
         return true;
     }
     
-	private void matched(HttpMessage msg, int id) {
+	private void matched(Matcher matcher, HttpMessage msg, int id) {
+		String tag = getConf();
 		if (tagHistoryType(msg.getHistoryRef().getHistoryType())) {
-			parent.addTag(id, this.getConf());
+			if (matcher.groupCount() > 0) {
+				StringBuffer result = new StringBuffer();
+				matcher.appendReplacement(result, tag);
+				tag = result.substring(matcher.start());
```

Comment:
Alternative could be
`tag = matcher.group().replaceAll(matcher.pattern().pattern(), tag);`

Reply:
Yeah, that's probably better, it avoids allocating a (potential) bigger string. Though it could avoid compiling the pattern again and replace first (there's just one to replace):
`matcher.pattern().matcher(matcher.group()).replaceFirst(tag);`

Reply:
K I’ll tweak that